<a href="https://colab.research.google.com/github/Decadent-tech/Exploratory-Data-Analysis/blob/main/IMDB_Movie_Business_data_exploration.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!gdown 1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd
!gdown 1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm

Downloading...
From: https://drive.google.com/uc?id=1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd
To: /content/movies.csv
100% 112k/112k [00:00<00:00, 30.6MB/s]
Downloading...
From: https://drive.google.com/uc?id=1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm
To: /content/directors.csv
100% 65.4k/65.4k [00:00<00:00, 43.0MB/s]


In [3]:
import pandas as pd
import numpy as np

In [4]:
movies = pd.read_csv('movies.csv')
movies.head()

Unnamed: 0     id     budget  popularity     revenue  \
0           0  43597  237000000         150  2787965087   
1           1  43598  300000000         139   961000000   
2           2  43599  245000000         107   880674609   
3           3  43600  250000000         112  1084939099   
4           5  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday

**So what kind of questions can we ask from this dataset?**

- **Top 10 most popular movies**, using `popularity`.
- Find the **highest rated movies**, using `vote_average`.
- We can find number of **movies released per year**.
- Find **highest budget movies in a year** using both `budget` and `year`.

**But can we ask more interesting/deeper questions?**

- Do you think we can find the **most productive directors**?
- Which **directors produce high budget films**?
- **Highest and lowest rated movies for every month** in a particular year?

index_col=0 explicitly states to treat the first column as the index.

The default value is index_col=None

In [5]:
movies = pd.read_csv('movies.csv', index_col=0)
movies.head()

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
5  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
5                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
5         4767  2007   May   Tuesday

The movies dataframe contains 1465 rows and 11 columns

In [6]:
movies.shape

(1465, 11)

In [7]:
directors = pd.read_csv('directors.csv', index_col=0)
directors.head()

director_name    id gender
0      James Cameron  4762   Male
1     Gore Verbinski  4763   Male
2         Sam Mendes  4764   Male
3  Christopher Nolan  4765   Male
4     Andrew Stanton  4766   Male

The `directors` dataframe contains 2349 rows and 3 columns.

In [8]:
directors.shape

(2349, 3)

If you observe,

- `director_id` of movies are taken from `id` of directors.
- Thus, we can merge our dataframes based on these two columns as **keys**.

In [9]:
movies['director_id'].nunique()

199

In [10]:
directors['id'].nunique()

2349

**Summary:**

- `movies` dataset: 1465 rows, but only 199 unique directors
- `directors` dataset: 2349 unique directors (equal to the no. of rows)

**What can we infer from this?**

- The directors in `movies` data is a subset of directors in `directors` data.

**How can we check if all `director_id` values are present in `id`?**

In [11]:
movies['director_id'].isin(directors['id'])

0       True
1       True
2       True
3       True
5       True
        ... 
4736    True
4743    True
4748    True
4749    True
4768    True
Name: director_id, Length: 1465, dtype: bool

**How can we check if there's any False here?**

In [13]:
np.all(movies['director_id'].isin(directors['id']))

True

- Only the ones for which we have a corresponding row in `movies`.

**So which `join` type do you think we should apply here?**

- `LEFT` Join

In [14]:
data = movies.merge(directors, how='left', left_on='director_id',right_on='id')
data

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      director_id  year month       day      director_name  id_y gender  
0            4762  2009   Dec  Thursday      James Cameron  4762   Male  
1            4763  2007   May  Saturday     Gore Verbinski  4763   Male  
2            4764  2015   Oct    Monday         Sam Mendes  4764   Male  
3            4765  2012   Jul    Monday  Christopher Nolan  4765   Male  
4            4767  2007   May   Tuesday          Sam Raimi  4767   Male  
...           ...   ...   ...       ...                ...   ...    ...  
1460         4809  1978   May    Monday    Martin Scorsese  4809   Male  
1461         5369  1994   Sep   Tuesday        Kevin Smith  5369   Male  
1462         5148  2009   Aug    Friday           Uwe Boll  5148   Male  
1463         5535  1990   Jul    Friday  Richard Linklater  5535   Male  
1464         5097  1992   Sep    Friday   Robert Rodriguez  5097    NaN  

[1465 rows x 14 columns]

In [15]:
data.head()

id_x     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
4  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day      director_name  id_y gender  
0         4762  2009   Dec  Thursday      James Cameron  4762   Male  
1         4763  2007   May  Saturday     Gore Verbinski  4763   Male  
2         4764  2015   Oct    Monday         Sam Mendes  4764   Male  
3         4765  2012   Jul    Monday  Christopher Nolan  4765   Male  
4         4767  2007   May   Tuesday          Sam Raimi  4767   Male

In [16]:
data.drop(['director_id','id_y'], axis=1, inplace=True)#dropping two redundant columns
data.head()

id_x     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
4  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   
2                                   Spectre           6.3        4466  2015   
3                     The Dark Knight Rises           7.6        9106  2012   
4                              Spider-Man 3           5.9        3576  2007   

  month       day      director_name gender  
0   Dec  Thursday      James Cameron   Male  
1   May  Saturday     Gore Verbinski   Male  
2   Oct    Monday         Sam Mendes   Male  
3   Jul    Monday  Christopher Nolan   Male  
4   May   Tuesday          Sam Raimi   Male

In [17]:
def encode(data):
  if data == "Male":
    return 0
  else:
    return 1

In [18]:
data['gender'] = data['gender'].apply(encode)
data

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name  gender  
0     2009   Dec  Thursday      James Cameron       0  
1     2007   May  Saturday     Gore Verbinski       0  
2     2015   Oct    Monday         Sam Mendes       0  
3     2012   Jul    Monday  Christopher Nolan       0  
4     2007   May   Tuesday          Sam Raimi       0  
...    ...   ...       ...                ...     ...  
1460  1978   May    Monday    Martin Scorsese       0  
1461  1994   Sep   Tuesday        Kevin Smith       0  
1462  2009   Aug    Friday           Uwe Boll       0  
1463  1990   Jul    Friday  Richard Linklater       0  
1464  1992   Sep    Friday   Robert Rodriguez       1  

[1465 rows x 12 columns]

In [19]:
#axis=0  →  It will apply to each column
data[['revenue', 'budget']].apply(np.sum)#in line every element gets looped same like numpy vectorization

revenue    209866997305
budget      70353617179
dtype: int64

In [20]:
data[['revenue', 'budget']].apply(np.sum, axis=1)#By setting the axis=1, every row of revenue was added to same row of budget.

0       3024965087
1       1261000000
2       1125674609
3       1334939099
4       1148871626
           ...    
1460        321952
1461       3178130
1462             0
1463             0
1464       2260920
Length: 1465, dtype: int64

In [21]:
def prof(x):
  return x['revenue'] - x['budget']

data['profit']=data[['revenue', 'budget']].apply(prof,axis=1)
data.head()

id_x     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
4  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   
2                                   Spectre           6.3        4466  2015   
3                     The Dark Knight Rises           7.6        9106  2012   
4                              Spider-Man 3           5.9        3576  2007   

  month       day      director_name  gender      profit  
0   Dec  Thursday      James Cameron       0  2550965087  
1   May  Saturday     Gore Verbinski       0   661000000  
2   Oct    Monday         Sam Mendes       0   635674609  
3   Jul    Monday  Christopher Nolan       0   834939099  
4   May   Tuesday          Sam Raimi       0   632871626

In [22]:
data.groupby('director_name')

In [23]:
data.groupby('director_name').ngroups

199

In [24]:
data.groupby('director_name').groups

{'Adam McKay': [176, 323, 366, 505, 839, 916], 'Adam Shankman': [265, 300, 350, 404, 458, 843, 999, 1231], 'Alejandro González Iñárritu': [106, 749, 1015, 1034, 1077, 1405], 'Alex Proyas': [95, 159, 514, 671, 873], 'Alexander Payne': [793, 1006, 1101, 1211, 1281], 'Andrew Adamson': [11, 43, 328, 501, 947], 'Andrew Niccol': [533, 603, 701, 722, 1439], 'Andrzej Bartkowiak': [349, 549, 754, 911, 924], 'Andy Fickman': [517, 681, 909, 926, 973, 1023], 'Andy Tennant': [314, 320, 464, 593, 676, 885], 'Ang Lee': [99, 134, 748, 840, 1089, 1110, 1132, 1184], 'Anne Fletcher': [610, 650, 736, 789, 1206], 'Antoine Fuqua': [310, 338, 424, 467, 576, 808, 818, 1105], 'Atom Egoyan': [946, 1128, 1164, 1194, 1347, 1416], 'Barry Levinson': [313, 319, 471, 594, 878, 898, 1013, 1037, 1082, 1143, 1185, 1345, 1378], 'Barry Sonnenfeld': [13, 48, 90, 205, 591, 778, 783], 'Ben Stiller': [209, 212, 547, 562, 850], 'Bill Condon': [102, 307, 902, 1233, 1381], 'Bobby Farrelly': [352, 356, 481, 498, 624, 630, 654, 80

In [25]:
data.groupby('director_name').get_group('Alexander Payne')

id_x    budget  popularity    revenue            title  vote_average  \
793   45163  30000000          19  105834556    About Schmidt           6.7   
1006  45699  20000000          40  177243185  The Descendants           6.7   
1101  46004  16000000          23  109502303         Sideways           6.9   
1211  46446  12000000          29   17654912         Nebraska           7.4   
1281  46813         0          13          0         Election           6.7   

      vote_count  year month       day    director_name  gender     profit  
793          362  2002   Dec    Friday  Alexander Payne       1   75834556  
1006         934  2011   Sep    Friday  Alexander Payne       1  157243185  
1101         478  2004   Oct    Friday  Alexander Payne       1   93502303  
1211         636  2013   Sep  Saturday  Alexander Payne       1    5654912  
1281         270  1999   Apr    Friday  Alexander Payne       1          0

**How can we find the count of movies by each director?**

In [26]:
data.groupby('director_name')['title'].count()

director_name
Adam McKay                      6
Adam Shankman                   8
Alejandro González Iñárritu     6
Alex Proyas                     5
Alexander Payne                 5
                               ..
Wes Craven                     10
Wolfgang Petersen               7
Woody Allen                    18
Zack Snyder                     7
Zhang Yimou                     6
Name: title, Length: 199, dtype: int64

**How to find multiple aggregates of any feature?**

Finding the very first year and the latest year a director released a movie i.e basically the **min** & **max** of the `year` column, grouped by `director_name`.

In [27]:
data.groupby(['director_name'])["year"].aggregate(['min', 'max'])

min   max
director_name                          
Adam McKay                   2004  2015
Adam Shankman                2001  2012
Alejandro González Iñárritu  2000  2015
Alex Proyas                  1994  2016
Alexander Payne              1999  2013
...                           ...   ...
Wes Craven                   1984  2011
Wolfgang Petersen            1981  2006
Woody Allen                  1977  2013
Zack Snyder                  2004  2016
Zhang Yimou                  2002  2014

[199 rows x 2 columns]

1. We can **group** the data by director and use `max` of the budget as aggregator.

In [28]:
data_dir_budget = data.groupby("director_name")["budget"].max().reset_index()
data_dir_budget.head()

director_name     budget
0                   Adam McKay  100000000
1                Adam Shankman   80000000
2  Alejandro González Iñárritu  135000000
3                  Alex Proyas  140000000
4              Alexander Payne   30000000

2. We can **filter** out the director names with **max budget >100M**.

In [29]:
names = data_dir_budget.loc[data_dir_budget["budget"] >= 100, "director_name"]

3. Finally, we can filter out the details of the movies by these directors.

In [30]:
data.loc[data['director_name'].isin(names)]

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name  gender      profit  
0     2009   Dec  Thursday      James Cameron       0  2550965087  
1     2007   May  Saturday     Gore Verbinski       0   661000000  
2     2015   Oct    Monday         Sam Mendes       0   635674609  
3     2012   Jul    Monday  Christopher Nolan       0   834939099  
4     2007   May   Tuesday          Sam Raimi       0   632871626  
...    ...   ...       ...                ...     ...         ...  
1460  1978   May    Monday    Martin Scorsese       0      321952  
1461  1994   Sep   Tuesday        Kevin Smith       0     3124130  
1462  2009   Aug    Friday           Uwe Boll       0           0  
1463  1990   Jul    Friday  Richard Linklater       0           0  
1464  1992   Sep    Friday   Robert Rodriguez       1     1820920  

[1465 rows x 13 columns]

**Can we filter groups in a single go using Lambda functions?** Yes!

In [31]:
data.groupby('director_name').filter(lambda x: x["budget"].max() >= 100)

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name  gender      profit  
0     2009   Dec  Thursday      James Cameron       0  2550965087  
1     2007   May  Saturday     Gore Verbinski       0   661000000  
2     2015   Oct    Monday         Sam Mendes       0   635674609  
3     2012   Jul    Monday  Christopher Nolan       0   834939099  
4     2007   May   Tuesday          Sam Raimi       0   632871626  
...    ...   ...       ...                ...     ...         ...  
1460  1978   May    Monday    Martin Scorsese       0      321952  
1461  1994   Sep   Tuesday        Kevin Smith       0     3124130  
1462  2009   Aug    Friday           Uwe Boll       0           0  
1463  1990   Jul    Friday  Richard Linklater       0           0  
1464  1992   Sep    Friday   Robert Rodriguez       1     1820920  

[1465 rows x 13 columns]

# we want to filter the risky movies whose budget was even higher than the average revenue of the director from his other movies.

In [32]:
def func(x):
  # returns whether a movie is risky or not
  x["risky"] = x["budget"] - x["revenue"].mean() >= 0
  return x

data_risky = data.groupby("director_name", group_keys=False).apply(func)
data_risky

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name  gender      profit  risky  
0     2009   Dec  Thursday      James Cameron       0  2550965087  False  
1     2007   May  Saturday     Gore Verbinski       0   661000000  False  
2     2015   Oct    Monday         Sam Mendes       0   635674609  False  
3     2012   Jul    Monday  Christopher Nolan       0   834939099  False  
4     2007   May   Tuesday          Sam Raimi       0   632871626  False  
...    ...   ...       ...                ...     ...         ...    ...  
1460  1978   May    Monday    Martin Scorsese       0      321952  False  
1461  1994   Sep   Tuesday        Kevin Smith       0     3124130  False  
1462  2009   Aug    Friday           Uwe Boll       0           0  False  
1463  1990   Jul    Friday  Richard Linklater       0           0  False  
1464  1992   Sep    Friday   Robert Rodriguez       1     1820920  False  

[1465 rows x 14 columns]

In [33]:
data_risky.loc[data_risky["risky"]]

id_x     budget  popularity     revenue  \
7     43608  200000000         107   586090727   
12    43614  380000000         135  1045713802   
15    43618  200000000          37   310669540   
20    43624  209000000          64   303025485   
24    43630  210000000           3   459359555   
...     ...        ...         ...         ...   
1347  47224    5000000           7     3263585   
1349  47229    5000000           3     4842699   
1351  47233    5000000           6           0   
1356  47263   15000000          10           0   
1383  47453    3500000           4           0   

                                            title  vote_average  vote_count  \
7                               Quantum of Solace           6.1        2965   
12    Pirates of the Caribbean: On Stranger Tides           6.4        4948   
15                                     Robin Hood           6.2        1398   
20                                     Battleship           5.5        2114   
24                          X-Men: The Last Stand           6.3        3525   
...                                           ...           ...         ...   
1347                          The Sweet Hereafter           6.8         103   
1349                         90 Minutes in Heaven           5.4          40   
1351                                Light Sleeper           5.7          15   
1356                           Dying of the Light           4.5         118   
1383                  In the Name of the King III           3.3          19   

      year month        day   director_name  gender     profit  risky  
7     2008   Oct   Thursday    Marc Forster       0  386090727   True  
12    2011   May   Saturday    Rob Marshall       0  665713802   True  
15    2010   May  Wednesday    Ridley Scott       0  110669540   True  
20    2012   Apr  Wednesday      Peter Berg       0   94025485   True  
24    2006   May  Wednesday    Brett Ratner       0  249359555   True  
...    ...   ...        ...             ...     ...        ...    ...  
1347  1997   May  Wednesday     Atom Egoyan       0   -1736415   True  
1349  2015   Sep     Friday  Michael Polish       0    -157301   True  
1351  1992   Aug     Friday   Paul Schrader       1   -5000000   True  
1356  2014   Dec   Thursday   Paul Schrader       1  -15000000   True  
1383  2013   Dec     Friday        Uwe Boll       0   -3500000   True  

[131 rows x 14 columns]

Similarly, we can do it for the `budget` as well.

In [34]:
data['budget']=(data['budget']/1000000).round(2)
data.head()

id_x  budget  popularity     revenue  \
0  43597   237.0         150  2787965087   
1  43598   300.0         139   961000000   
2  43599   245.0         107   880674609   
3  43600   250.0         112  1084939099   
4  43602   258.0         115   890871626   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   
2                                   Spectre           6.3        4466  2015   
3                     The Dark Knight Rises           7.6        9106  2012   
4                              Spider-Man 3           5.9        3576  2007   

  month       day      director_name  gender      profit  
0   Dec  Thursday      James Cameron       0  2550965087  
1   May  Saturday     Gore Verbinski       0   661000000  
2   Oct    Monday         Sam Mendes       0   635674609  
3   Jul    Monday  Christopher Nolan       0   834939099  
4   May   Tuesday          Sam Raimi       0   632871626

 **Let's say we are interested in fetching all the highly rated movies    movies with ratings > 7**

In [36]:
data.loc[data['vote_average'] > 7]

id_x  budget  popularity     revenue  \
0     43597  237.00         150  2787965087   
3     43600  250.00         112  1084939099   
14    43616  250.00         120   956019788   
16    43619  250.00          94   958400000   
19    43622  200.00         100  1845034188   
...     ...     ...         ...         ...   
1456  48321    0.01          20     7000000   
1457  48323    0.00           5           0   
1458  48335    0.06          27     3221152   
1460  48363    0.00           3      321952   
1461  48370    0.03          19     3151130   

                                          title  vote_average  vote_count  \
0                                        Avatar           7.2       11800   
3                         The Dark Knight Rises           7.6        9106   
14    The Hobbit: The Battle of the Five Armies           7.1        4760   
16          The Hobbit: The Desolation of Smaug           7.6        4524   
19                                      Titanic           7.5        7562   
...                                         ...           ...         ...   
1456                                 Eraserhead           7.5         485   
1457                                 The Mighty           7.1          51   
1458                                         Pi           7.1         586   
1460                             The Last Waltz           7.9          64   
1461                                     Clerks           7.4         755   

      year month        day      director_name  gender      profit  
0     2009   Dec   Thursday      James Cameron       0  2550965087  
3     2012   Jul     Monday  Christopher Nolan       0   834939099  
14    2014   Dec  Wednesday      Peter Jackson       0   706019788  
16    2013   Dec  Wednesday      Peter Jackson       0   708400000  
19    1997   Nov    Tuesday      James Cameron       0  1645034188  
...    ...   ...        ...                ...     ...         ...  
1456  1977   Mar   Saturday        David Lynch       0     6990000  
1457  1998   Oct     Friday      Peter Chelsom       0           0  
1458  1998   Jul     Friday   Darren Aronofsky       0     3161152  
1460  1978   May     Monday    Martin Scorsese       0      321952  
1461  1994   Sep    Tuesday        Kevin Smith       0     3124130  

[301 rows x 13 columns]

You can also perform the filtering without using `loc`.




In [37]:
data[data['vote_average'] > 7]

id_x  budget  popularity     revenue  \
0     43597  237.00         150  2787965087   
3     43600  250.00         112  1084939099   
14    43616  250.00         120   956019788   
16    43619  250.00          94   958400000   
19    43622  200.00         100  1845034188   
...     ...     ...         ...         ...   
1456  48321    0.01          20     7000000   
1457  48323    0.00           5           0   
1458  48335    0.06          27     3221152   
1460  48363    0.00           3      321952   
1461  48370    0.03          19     3151130   

                                          title  vote_average  vote_count  \
0                                        Avatar           7.2       11800   
3                         The Dark Knight Rises           7.6        9106   
14    The Hobbit: The Battle of the Five Armies           7.1        4760   
16          The Hobbit: The Desolation of Smaug           7.6        4524   
19                                      Titanic           7.5        7562   
...                                         ...           ...         ...   
1456                                 Eraserhead           7.5         485   
1457                                 The Mighty           7.1          51   
1458                                         Pi           7.1         586   
1460                             The Last Waltz           7.9          64   
1461                                     Clerks           7.4         755   

      year month        day      director_name  gender      profit  
0     2009   Dec   Thursday      James Cameron       0  2550965087  
3     2012   Jul     Monday  Christopher Nolan       0   834939099  
14    2014   Dec  Wednesday      Peter Jackson       0   706019788  
16    2013   Dec  Wednesday      Peter Jackson       0   708400000  
19    1997   Nov    Tuesday      James Cameron       0  1645034188  
...    ...   ...        ...                ...     ...         ...  
1456  1977   Mar   Saturday        David Lynch       0     6990000  
1457  1998   Oct     Friday      Peter Chelsom       0           0  
1458  1998   Jul     Friday   Darren Aronofsky       0     3161152  
1460  1978   May     Monday    Martin Scorsese       0      321952  
1461  1994   Sep    Tuesday        Kevin Smith       0     3124130  

[301 rows x 13 columns]

In [39]:
data.loc[data['vote_average'] > 7,['title','director_name']]

title      director_name
0                                        Avatar      James Cameron
3                         The Dark Knight Rises  Christopher Nolan
14    The Hobbit: The Battle of the Five Armies      Peter Jackson
16          The Hobbit: The Desolation of Smaug      Peter Jackson
19                                      Titanic      James Cameron
...                                         ...                ...
1456                                 Eraserhead        David Lynch
1457                                 The Mighty      Peter Chelsom
1458                                         Pi   Darren Aronofsky
1460                             The Last Waltz    Martin Scorsese
1461                                     Clerks        Kevin Smith

[301 rows x 2 columns]

So far, we've only seen single condition based filtering.

#### What if we want to filter highly rated movies released after 2014?

Notice that two different conditions are involved here.

1. Movies should be highly rated i.e. ratings > 7
2. Movies should be released either in the year 2015 or later.

We can use the `&` operator to combine multiple conditions.

In [40]:
data.loc[(data['vote_average'] > 7) & (data['year'] >= 2015)].head()

id_x  budget  popularity     revenue                    title  \
30   43641   190.0         102  1506249360                Furious 7   
78   43724   150.0         434   378858340       Mad Max: Fury Road   
106  43773   135.0         100   532950503             The Revenant   
162  43867   108.0         167   630161890              The Martian   
312  44128    75.0          48   108145109  The Man from U.N.C.L.E.   

     vote_average  vote_count  year month        day  \
30            7.3        4176  2015   Apr  Wednesday   
78            7.2        9427  2015   May  Wednesday   
106           7.3        6396  2015   Dec     Friday   
162           7.6        7268  2015   Sep  Wednesday   
312           7.1        2265  2015   Aug   Thursday   

                   director_name  gender      profit  
30                     James Wan       0  1316249360  
78                 George Miller       0   228858340  
106  Alejandro González Iñárritu       0   397950503  
162                 Ridley Scott       0   522161890  
312                  Guy Ritchie       0    33145109

#### How can we find movies released on either Friday or Sunday?

In [41]:
data.loc[(data['day'] == 'Friday') | (data['day'] == 'Saturday')].head()

id_x  budget  popularity     revenue  \
1   43598   300.0         139   961000000   
12  43614   380.0         135  1045713802   
22  43627   200.0          35   783766341   
25  43632   150.0          21   836297228   
40  43656   200.0          45   769653595   

                                          title  vote_average  vote_count  \
1      Pirates of the Caribbean: At World's End           6.9        4500   
12  Pirates of the Caribbean: On Stranger Tides           6.4        4948   
22                                 Spider-Man 2           6.7        4321   
25          Transformers: Revenge of the Fallen           6.0        3138   
40                                         2012           5.6        4903   

    year month       day    director_name  gender     profit  
1   2007   May  Saturday   Gore Verbinski       0  661000000  
12  2011   May  Saturday     Rob Marshall       0  665713802  
22  2004   Jun    Friday        Sam Raimi       0  583766341  
25  2009   Jun    Friday      Michael Bay       0  686297228  
40  2009   Oct  Saturday  Roland Emmerich       0  569653595

#  Top 5 most popular movies?

In [42]:
data.sort_values(['popularity'],ascending=False).head(5)

id_x  budget  popularity     revenue  \
58   43692   165.0         724   675120017   
78   43724   150.0         434   378858340   
119  43796   140.0         271   655011224   
120  43797   125.0         206   752100229   
45   43662   185.0         187  1004558444   

                                                 title  vote_average  \
58                                        Interstellar           8.1   
78                                  Mad Max: Fury Road           7.2   
119  Pirates of the Caribbean: The Curse of the Bla...           7.5   
120              The Hunger Games: Mockingjay - Part 1           6.6   
45                                     The Dark Knight           8.2   

     vote_count  year month        day      director_name  gender     profit  
58        10867  2014   Nov  Wednesday  Christopher Nolan       0  510120017  
78         9427  2015   May  Wednesday      George Miller       0  228858340  
119        6985  2003   Jul  Wednesday     Gore Verbinski       0  515011224  
120        5584  2014   Nov    Tuesday   Francis Lawrence       0  627100229  
45        12002  2008   Jul  Wednesday  Christopher Nolan       0  819558444

On applying this to a string column, it sorts the dataframe ***lexicographically**.

In [43]:
data.sort_values(['title'],ascending=False).head(5)

id_x  budget  popularity    revenue                    title  \
436  44364    60.0          36   71073932  xXx: State of the Union   
330  44165    70.0          46  277448382                      xXx   
994  45681    15.0          21    2856712                 eXistenZ   
547  44594    50.0          37   55969000              Zoolander 2   
850  45313    28.0          38   60780981                Zoolander   

     vote_average  vote_count  year month        day     director_name  \
436           4.7         549  2005   Apr  Wednesday      Lee Tamahori   
330           5.8        1424  2002   Aug     Friday         Rob Cohen   
994           6.7         475  1999   Apr  Wednesday  David Cronenberg   
547           4.7         797  2016   Feb   Saturday       Ben Stiller   
850           6.1        1337  2001   Sep     Friday       Ben Stiller   

     gender     profit  
436       0   11073932  
330       0  207448382  
994       0  -12143288  
547       0    5969000  
850       0   32780981

#### How will get list of movies directed by a particular director, say 'Christopher Nolan'?

In [44]:
data.loc[data['director_name'] == 'Christopher Nolan',['title']]

title
3     The Dark Knight Rises
45          The Dark Knight
58             Interstellar
59                Inception
74            Batman Begins
565                Insomnia
641            The Prestige
1341                Memento